cocotb/bfmgen.py

```diff
@@ -0,0 +1,364 @@
+    
+    with open(args.o, "w") as out:
+        out.write("//***************************************************************************\n")
+        out.write("//* BFMs file for cocotb. \n")
+        out.write("//* Note: This file is generated by cocotb-bfmgen. Do Not Edit\n")
+        out.write("//***************************************************************************\n")
+
+        for t in inst.bfm_type_info_m.keys():
+            info = inst.bfm_type_info_m[t]
+        
+            if cocotb.bfm_vlog not in info.hdl.keys():
+                raise Exception("BFM \"" + t.__name__ + "\" does not support Verilog")
+        
+            with open(info.hdl[cocotb.bfm_vlog], "r") as template_f:
+                template = template_f.read()
+        
+            out.write(process_template_vl(template, info))
+        
+def process_template_sv(template, bfm_name, info):
+        
+    t = Template(template)
+    
+    bfm_import_calls = ""
+    for i in range(len(info.import_info)):
+        imp = info.import_info[i]
+        bfm_import_calls += "              " + str(i) + ": begin\n"
+        # Verilator doesn't evaluate expressions in the order that
+        # they appear in the argument list. Consequently, we need 
+        # to create temporary variables to ensure the order is correct.
+        for pi in range(len(imp.signature)):
+            p = imp.signature[pi]
+            if p.ptype.s:
+                bfm_import_calls += "                  longint p" + str(pi) + " = cocotb_bfm_get_si_param(bfm_id);\n"
+            else:
+                bfm_import_calls += "                  longint unsigned p" + str(pi) + " = cocotb_bfm_get_ui_param(bfm_id);\n"
+                
+        bfm_import_calls += "                  " + imp.T.__name__  + "(\n"
+        for pi in range(len(imp.signature)):
+            bfm_import_calls += "                      p" + str(pi)
+
+            if pi+1 < len(imp.signature):
+                bfm_import_calls += ","
+            bfm_import_calls += "\n"
+        bfm_import_calls += "                      );\n"
+        bfm_import_calls += "              end\n"
+        
+    bfm_export_tasks = ""
+    for i in range(len(info.export_info)):
+        exp = info.export_info[i]
+        bfm_export_tasks += "    task " + exp.T.__name__ + "("
+        for j in range(len(exp.signature)):
+            p = exp.signature[j]
+            bfm_export_tasks += p.ptype.vl_type() + " " + p.pname
+            if j+1 < len(exp.signature):
+                bfm_export_tasks += ", "
+        bfm_export_tasks += ");\n"
+        bfm_export_tasks += "    begin\n"
+        bfm_export_tasks += "        cocotb_bfm_begin_msg(bfm_id, " + str(i) + ");\n"
+        for p in exp.signature:
+            if p.ptype.s:
+                bfm_export_tasks += "        cocotb_bfm_add_si_param(bfm_id, " + p.pname + ");\n"
+            else:
+                bfm_export_tasks += "        cocotb_bfm_add_ui_param(bfm_id, " + p.pname + ");\n"
```

Comment:
nit: tempting to write these using `.format`, such as
```suggestion
                bfm_export_tasks += "        cocotb_bfm_add_ui_param(bfm_id, {});\n".format(p.pname)
```